### PR TITLE
doc: sync data structures with pybind11 v1

### PR DESCRIPTION
**Pull request by james-d-mitchell**

### CODE REVIEW

File: include/libsemigroups/words.hpp [hunk 1]
```diff
@@ -56,11 +56,32 @@ namespace libsemigroups {
...
+  //!  - \ref Strings
+  //!  - \ref random_word
+  //!
+  //! * counting words:
+  //!
+  //!   - \ref number_of_words
+  //!
+  //! * converting to and from strings and words:
+  //!
+  //!   - \ref to_string
```
Comment (Joseph-Edwards) [inline include/libsemigroups/words.hpp:76]: This link takes you to the `to_string` function that makes a string representation of an `AhoCorasick`. I think you can specify the function you want by specifying the relevant argument types.

File: include/libsemigroups/words.hpp [hunk 2]
```diff
@@ -56,11 +56,32 @@ namespace libsemigroups {
...
+  //!   - \ref to_word
```
Comment (Joseph-Edwards) [inline include/libsemigroups/words.hpp:77]: This, for some reason, doesn't create a link

File: include/libsemigroups/paths.hpp [hunk 3]
```diff
@@ -921,8 +956,7 @@ namespace libsemigroups {
    public:
     output_type get() const {
       set_iterator();
-      return std::visit(
-          [](auto& it) -> auto const& { return *it; }, _current);
+      return std::visit([](auto& it) -> auto const& { return *it; }, _current);
```
Comment (Joseph-Edwards) [inline include/libsemigroups/paths.hpp:959]:
```suggestion
      return std::visit(
          [](auto& it) -> auto const& { return *it; }, _current);
```
For clang-format

File: include/libsemigroups/words.hpp [hunk 4]
```diff
@@ -438,8 +459,7 @@ namespace libsemigroups {
     //! function could be anything.
     [[nodiscard]] output_type get() const noexcept {
       set_iterator();
-      return std::visit(
-          [](auto& it) -> auto const& { return *it; }, _current);
+      return std::visit([](auto& it) -> auto const& { return *it; }, _current);
```
Comment (Joseph-Edwards) [inline include/libsemigroups/words.hpp:462]:
```suggestion
      return std::visit(
          [](auto& it) -> auto const& { return *it; }, _current);
```
For clang-format

**Review by Joseph-Edwards — CHANGES_REQUESTED** [review]
Other than the broken links and the formatting mentioned above, this looks good to me!

File: include/libsemigroups/words.hpp (hunk 1 above)
Comment (james-d-mitchell) [reply include/libsemigroups/words.hpp]: Ah good catch, I've fixed this!

File: include/libsemigroups/words.hpp (hunk 2 above)
Comment (james-d-mitchell) [reply include/libsemigroups/words.hpp]: Fixed!

File: include/libsemigroups/paths.hpp (hunk 3 above)
Comment (james-d-mitchell) [reply include/libsemigroups/paths.hpp]: Fixed

File: include/libsemigroups/words.hpp (hunk 4 above)
Comment (james-d-mitchell) [reply include/libsemigroups/words.hpp]: Fixed

**Pull request merged**